- Nama :  ....
- Batch : ....
- Objective : Menguji kemampuan pada SQL, pandas dan python.

# P0 LC2 Exercise

- bigquery project_id: **`bigquery-public-data`**
- dataset: **`austin_bikeshare`**
- tabel: **`bikeshare_stations`**

# Setup

In [1]:
from google.colab import auth
from google.cloud import bigquery
import pandas as pd

auth.authenticate_user()

project_id = "ayusudi-e52a2" # <- your project id
client = bigquery.Client(project=project_id)

# Soal

## SQL

Jawab masing-masing pertanyaan ini dengan 1 query:

1. Apa saja nama station yang memiliki status `active`?
2. Berapa banyak station yang berada di `council_district = 9` dan memiliki status `active`?
3. Bandingkan banyaknya station antara `power_type = solar` dengan `power type = non-metered`?

## Pandas

Jawab masing-masing pertanyaan dengan pandas:
1. Tampilkan summary data! Berikan insight terkait summary data yang ditemukan!
2. Buatkan kolom baru `keterangan` yang memiliki ketentuan berikut
  - Jika `number_of_docks` antara 0 - 5 -> value `station kecil`
  - Jika `number_of_docks` antara 6 - 10 -> value `station sedang`
  - Jika `number_of_docks` lebih dari/sama dengan 11 -> value `station besar`
3. Tampilkan nama station yang memiliki `power_type = solar` dan `council_district = 10`

## Sample Query

In [29]:
df = client.query('''
SELECT *
FROM `bigquery-public-data.austin_bikeshare.bikeshare_stations`
''').to_dataframe()

df.head()

station_id              name  status            address alternate_name  \
0        1003   8th & Guadalupe  closed  800 Guadalupe St.           None   
1        1006  Zilker Park West  closed  Barton Springs Rd           None   
2        1008      Nueces @ 3rd  closed         311 Nueces           None   
3        2500   Republic Square  closed   425 W 4th Street           None   
4        2536  Waller & 6th St.  closed     602 Waller St.           None   

   city_asset_number property_type  number_of_docks power_type  \
0               <NA>          None             <NA>       None   
1               <NA>          None             <NA>       None   
2               <NA>          None             <NA>       None   
3               <NA>          None             <NA>       None   
4               <NA>          None             <NA>       None   

   footprint_length  footprint_width notes  council_district  \
0              <NA>              NaN  None                 9   
1              <NA>              NaN  None                 8   
2              <NA>              NaN  None                 9   
3              <NA>              NaN  None                 9   
4              <NA>              NaN  None                 3   

              modified_date  
0 2021-01-04 12:00:00+00:00  
1 2021-01-04 12:00:00+00:00  
2 2021-01-04 12:00:00+00:00  
3 2021-01-04 12:00:00+00:00  
4 2021-01-04 12:00:00+00:00

# SQL

## 1. Apa saja nama station yang memiliki status `active`?

In [8]:
# code here
df = client.query('''
SELECT COUNT(*)
FROM `bigquery-public-data.austin_bikeshare.bikeshare_stations`
WHERE status = 'active'
''').to_dataframe()
df

f0_
0   78

Knowledge : Total active station is 78

## 2. Berapa banyak station yang berada di `council_district = 9` dan memiliki status `active`?

In [10]:
# code here
df = client.query('''
SELECT COUNT(*)
FROM `bigquery-public-data.austin_bikeshare.bikeshare_stations`
WHERE
  status = 'active'
  AND council_district = 9
''').to_dataframe()
df

f0_
0   47

Total active station at council district 9 is 47.

## 3. Bandingkan banyaknya station antara `power_type = solar` dengan `power type = non-metered`?

In [17]:
# code here
df = client.query('''
SELECT power_type, COUNT(*)
FROM `bigquery-public-data.austin_bikeshare.bikeshare_stations`
GROUP BY power_type
HAVING power_type = 'solar' OR power_type = 'non-metered'
''').to_dataframe()
df

power_type  f0_
0        solar   79
1  non-metered    3

Hasil query menyatakan power_type solar lebih banyak yaitu 79 sementara non-metered berjumlah 3.

# Pandas

## 1. Tampilkan summary data! Berikan insight terkait summary data yang ditemukan!

In [49]:
df = client.query('''
SELECT *
FROM `bigquery-public-data.austin_bikeshare.bikeshare_stations`
''').to_dataframe()
df

station_id                                      name  status  \
0          1003                           8th & Guadalupe  closed   
1          1006                          Zilker Park West  closed   
2          1008                              Nueces @ 3rd  closed   
3          2500                           Republic Square  closed   
4          2536                          Waller & 6th St.  closed   
..          ...                                       ...     ...   
97         2537                            West & 6th St.  active   
98         3377  MoPac Pedestrian Bridge @ Veterans Drive  active   
99         3794                    Dean Keeton & Speedway  active   
100        3795                      Dean Keeton & Whitis  active   
101        3797                         21st & University  active   

                    address alternate_name  city_asset_number  \
0         800 Guadalupe St.           None               <NA>   
1         Barton Springs Rd           None               <NA>   
2                311 Nueces           None               <NA>   
3          425 W 4th Street           None               <NA>   
4            602 Waller St.           None               <NA>   
..                      ...            ...                ...   
97            603 West Ave.           None              16737   
98        2105 Veterans Dr.           None              16598   
99   180 E. Dean Keeton St.           None              32670   
100  300 W. Dean Keeton St.           None              32671   
101    2007 University Ave.           None              32674   

            property_type  number_of_docks power_type  footprint_length  \
0                    None             <NA>       None              <NA>   
1                    None             <NA>       None              <NA>   
2                    None             <NA>       None              <NA>   
3                    None             <NA>       None              <NA>   
4                    None             <NA>       None              <NA>   
..                    ...              ...        ...               ...   
97   undetermined_parking               13      solar                40   
98   undetermined_parking               19      solar                55   
99   undetermined_parking               19      solar                55   
100  undetermined_parking               19      solar                55   
101  undetermined_parking               19      solar                55   

     footprint_width                     notes  council_district  \
0                NaN                      None                 9   
1                NaN                      None                 8   
2                NaN                      None                 9   
3                NaN                      None                 9   
4                NaN                      None                 3   
..               ...                       ...               ...   
97               5.0                      None                 9   
98               5.0  parkland in ROW/eastment                 9   
99               5.0                      None                 9   
100              5.0                      None                 9   
101              5.0    ROW might belong to UT                 9   

                modified_date  
0   2021-01-04 12:00:00+00:00  
1   2021-01-04 12:00:00+00:00  
2   2021-01-04 12:00:00+00:00  
3   2021-01-04 12:00:00+00:00  
4   2021-01-04 12:00:00+00:00  
..                        ...  
97  2021-01-04 12:00:00+00:00  
98  2021-01-04 12:00:00+00:00  
99  2021-01-04 12:00:00+00:00  
100 2021-01-04 12:00:00+00:00  
101 2021-01-04 12:00:00+00:00  

[102 rows x 14 columns]

Total data pada dataset ada 102 dengan data column yang akan digunakan untuk menjawab soal ini adalah number_of_docks yang merupakan integer.

In [51]:
df.dropna(subset=['number_of_docks'], inplace=True)

## 2. Buatkan kolom baru `keterangan` yang memiliki ketentuan berikut

- Jika `number_of_docks` antara 0 - 5 -> value `station kecil`
- Jika `number_of_docks` antara 6 - 10 -> value `station sedang`
- Jika `number_of_docks` lebih dari/sama dengan 11 -> value `station besar`

In [57]:
def kategoriStation(n):
  if n >= 0 and n <= 5 :
    return 'kecil'
  elif n >= 6 and n <= 10:
    return 'sedang'
  elif n >= 11:
    return 'besar'
  return '-'

keterangan = []
for data in df['number_of_docks'] :
  result = kategoriStation(data)
  keterangan.append(result)

df['keterangan'] = keterangan
df

station_id                                      name  status  \
20         3687                            Boardwalk West  active   
21         3684                   Congress & Cesar Chavez  active   
22         2572                       Barton Springs Pool  active   
23         2575                      Riverside @ S. Lamar  active   
24         2574                               Zilker Park  active   
..          ...                                       ...     ...   
97         2537                            West & 6th St.  active   
98         3377  MoPac Pedestrian Bridge @ Veterans Drive  active   
99         3794                    Dean Keeton & Speedway  active   
100        3795                      Dean Keeton & Whitis  active   
101        3797                         21st & University  active   

                                  address  \
20                   300 E. Riverside Dr.   
21                  124 W Cesar Chavez St   
22                2200 William Barton Dr.   
23   Presented by Austin Parks Foundation   
24                      Barton Springs Rd   
..                                    ...   
97                          603 West Ave.   
98                      2105 Veterans Dr.   
99                 180 E. Dean Keeton St.   
100                300 W. Dean Keeton St.   
101                  2007 University Ave.   

                                        alternate_name  city_asset_number  \
20                                                None              16683   
21                                                None              32594   
22                                                None              16758   
23                                                None              16736   
24   Zilker Park at Barton Springs and William Burt...              16759   
..                                                 ...                ...   
97                                                None              16737   
98                                                None              16598   
99                                                None              32670   
100                                               None              32671   
101                                               None              32674   

            property_type  number_of_docks power_type  footprint_length  \
20               parkland                9      solar                30   
21               parkland               11      solar                35   
22               parkland               13      solar                40   
23               parkland               15      solar                45   
24               parkland               17      solar                50   
..                    ...              ...        ...               ...   
97   undetermined_parking               13      solar                40   
98   undetermined_parking               19      solar                55   
99   undetermined_parking               19      solar                55   
100  undetermined_parking               19      solar                55   
101  undetermined_parking               19      solar                55   

     footprint_width                     notes  council_district  \
20               5.0                      None                 9   
21               5.0  parkland at ROW/easement                 9   
22               5.0                      None                 8   
23               5.0  parkland at ROW/easement                 5   
24               5.0  parkland at ROW/easement                 8   
..               ...                       ...               ...   
97               5.0                      None                 9   
98               5.0  parkland in ROW/eastment                 9   
99               5.0                      None                 9   
100              5.0                      None                 9   
101              5.0    ROW might belong to UT               

Insight:
Dari dataset terdapat data null atau kosong sehingga harus dibersihkan dahulu.
Total data yang diproses dengan fungsi kategoriStation adalah 82.

## 3. Tampilkan nama station yang memiliki `power_type = solar` dan `council_district = 10`

In [61]:
data3 = df[(df['power_type'] == 'solar') & (df['council_district'] == 10)]
data3['name']

43    Lake Austin Blvd @ Deep Eddy
92           Lake Austin & Enfield
Name: name, dtype: object

Insight: Ditemukan 2 data yaitu Lake Austin Blvd @ Deep Eddy dan Lake Austin & Enfield sebagai station dengan power type solar yang terletak di council district 10.